app_dart/lib/src/request_handlers/push_benchmark_to_center.dart

```diff
@@ -0,0 +1,124 @@
+import '../foundation/providers.dart';
+import '../model/appengine/time_series.dart';
+import '../model/appengine/time_series_value.dart';
+import '../request_handling/body.dart';
+import '../request_handling/request_handler.dart';
+import '../service/datastore.dart';
+
+/// Pushes devicelab benchmark results to Metrics Center storage.
+@immutable
+class PushBenchmarkToCenter extends RequestHandler<Body> {
+  PushBenchmarkToCenter(
+    Config config, {
+    @visibleForTesting Logging logger,
+    @visibleForTesting DatastoreService datastoreService,
+  })  : logger = logger ?? Providers.serviceScopeLogger(),
+        datastoreService =
+            datastoreService ?? DatastoreService.defaultProvider(config.db),
+        super(config: config);
+
+  final Logging logger;
+  final DatastoreService datastoreService;
+
+  @override
+  Future<Body> get() async {
+    final int sinceMillis = parseSinceMillis(request.uri.queryParameters);
+    if (sinceMillis == null) {
+      logger.warning('The sinceMillis parameter is missing');
+      return Body.empty;
+    }
+
+    final List<TimeSeriesValue> timeSeriesValues =
+        await readTimeSeriesValue(sinceMillis);
+    final Map<String, TimeSeries> timeSeriesMap = await readTimeSeriesMap();
+
+    final List<mc.MetricPoint> metricPoints =
+        await transform(timeSeriesMap, timeSeriesValues);
+    logger.debug('Transformed ${metricPoints.length} metric points.');
+    logger.debug('The first one is ${metricPoints.first}');
+
+    final mc.FlutterDestination destination =
+        await mc.FlutterDestination.makeFromCredentialsJson(
+            await config.metricsCenterServiceAccountJson);
+    await destination.update(metricPoints);
+    return Body.empty;
+  }
+
+  @visibleForTesting
+  int parseSinceMillis(Map<String, String> queryParameters) {
+    final String s = queryParameters['sinceMillis'] ?? '';
+    return int.tryParse(s);
+  }
+
+  /// Returns a map containing all time series keyed by its id.
+  Future<Map<String, TimeSeries>> readTimeSeriesMap() async {
+    final Map<String, TimeSeries> map = <String, TimeSeries>{};
+    await for (TimeSeries ts in datastoreService.db.query<TimeSeries>().run()) {
+      if (!(ts.id is String)) {
+        logger.error('The id of TimeSeries $ts is non-string: ${ts.id}');
+        continue;
+      }
+      map[ts.id as String] = ts;
+    }
+    return map;
+  }
+
+  /// Reads benchmark records with [createTimestamp] later than [sinceMillis].
+  /// And limits the result size to no larger than [batchSize].
+  Future<List<TimeSeriesValue>> readTimeSeriesValue(int sinceMillis,
+      {int batchSize = 1000}) async {
+    final Query<TimeSeriesValue> query =
+        datastoreService.db.query<TimeSeriesValue>()
+          ..filter('createTimestamp >', sinceMillis)
+          ..limit(batchSize)
+          ..order('createTimestamp');
+    return query.run().toList();
+  }
+
+  Future<List<mc.MetricPoint>> transform(Map<String, TimeSeries> timeSeriesMap,
```

Comment:
You can potentially move most of the content of this class to a service class and use it within the existing metrics handler to push the data. That way you won't need to synchronize timestamps and/or create an additional cronjob

Reply:
> The existing metrics handler only handles new data, @liyuqian and I would like to also transfer existing data to skia perf. Although we could export the existing data, however there's no easy way to process the exported data (leveldb log format).
> 
> Besides adding more burden to existing handler will increase its latency, complexity and decrease stability. For example, a failed pushing to metrics center could be caused by network interruptions temporarily, after the incident is over, there's no way to retry the pushing.

If that is the case then I'd recommend to split the process in two parts one for the data migration and the other one for just sending the data as it arrives.

Data migration will require different processes, maybe running for days, data validation on the target, running after work hours to minimize disruption, etc. 

For the second part the agent or the bots can call the two handlers with the same collected json.

Reply:
This PR is partially implementing the data migration process.

For the second part, do you mean letting cocoon agent calls `/api/push_benchmark_to_center` in addition to `/api/update-task-status`?

Reply:
BTW, @kf6gpe has a plan to have the new metrics co-exist in both SkiaPerf/MetricsCenter and Cocoon. That way, he can compare between two dashboards' regression alerts, and make sure that SkiaPerf's alerts are properly set before sunsetting the Cocoon dashboard.

Reply:
> This PR is partially implementing the data migration process.
> 
> For the second part, do you mean letting cocoon agent calls `/api/push_benchmark_to_center` in addition to `/api/update-task-status`?

Yes, that is correct.